pkg/cmd/image/inspect.go

```diff
@@ -19,58 +19,182 @@ package image
+	parsedIdentifier, err := referenceutil.ParseAnyReference(identifier)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("invalid identifier %s: %w", identifier, err)
+	}
+	digest := ""
+	if identifierDigest, hasDigest := parsedIdentifier.(reference.Digested); hasDigest {
+		digest = identifierDigest.Digest().String()
+	}
+	name := ""
+	if identifierName, hasName := parsedIdentifier.(reference.Named); hasName {
+		name = identifierName.Name()
+	}
+	tag := "latest"
+	if identifierTag, hasTag := parsedIdentifier.(reference.Tagged); hasTag && identifierTag.Tag() != "" {
+		tag = identifierTag.Tag()
+	}
+
+	// Initialize filters
+	var filters []string
+	// This will hold the final image list, if any
+	var imageList []images.Image
+
+	// No digest in the request? Then assume it is a name
+	if digest == "" {
+		filters = []string{fmt.Sprintf("name==%s:%s", name, tag)}
+		// Query it
+		imageList, err = client.ImageService().List(ctx, filters...)
+		if err != nil {
+			return nil, "", "", fmt.Errorf("containerd image service failed: %w", err)
+		}
+		// Nothing? Then it could be a short id (aka truncated digest) - we are going to use this
+		if len(imageList) == 0 {
+			digest = fmt.Sprintf("sha256:%s.*", regexp.QuoteMeta(strings.TrimPrefix(identifier, "sha256:")))
+			name = ""
+			tag = ""
+		} else {
+			// Otherwise, we found one by name. Get the digest from it.
+			digest = imageList[0].Target.Digest.String()
+		}
+	}
+
+	// At this point, we DO have a digest (or short id), so, that is what we are retrieving
+	filters = []string{fmt.Sprintf("target.digest~=^%s$", digest)}
+	imageList, err = client.ImageService().List(ctx, filters...)
+	if err != nil {
+		return nil, "", "", fmt.Errorf("containerd image service failed: %w", err)
+	}
+
+	// TODO: docker does allow retrieving images by Id, so implement as a last ditch effort (probably look-up the store)
+
+	// Return the list we found, along with normalized name and tag
+	return imageList, name, tag, nil
+}
+
 // Inspect prints detailed information of each image in `images`.
-func Inspect(ctx context.Context, client *containerd.Client, images []string, options types.ImageInspectOptions) error {
-	f := &imageInspector{
-		mode: options.Mode,
+func Inspect(ctx context.Context, client *containerd.Client, identifiers []string, options types.ImageInspectOptions) error {
+	// Verify we have a valid mode
+	// TODO: move this out of here, to Cobra command line arg validation
+	if options.Mode != "native" && options.Mode != "dockercompat" {
+		return fmt.Errorf("unknown mode %q", options.Mode)
 	}
-	walker := &imagewalker.ImageWalker{
-		Client: client,
-		OnFound: func(ctx context.Context, found imagewalker.Found) error {
-			ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
-			defer cancel()
 
-			n, err := imageinspector.Inspect(ctx, client, found.Image, options.GOptions.Snapshotter)
+	// Set a timeout
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
```

Comment:
Rather than hard-coded timeout, it should be cancellable from the user with a signal.

Reply:
The timeout was there already, hard-coded, with the same value.
I do agree we should do better, but maybe as a follow-up PR?

LMK your thoughts.

Reply:
OK, can be a following up.